realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncedRealmMigrationTests.java

```diff
@@ -56,4 +64,216 @@ public void migrateRealm_syncConfigurationThrows() {
+        dynamicRealm.close();
+
+        // Open typed Realm, which will validate the schema
+        Realm realm = Realm.getInstance(config);
+        RealmObjectSchema stringOnlySchema = realm.getSchema().get(className);
+        try {
+            assertTrue(stringOnlySchema.hasField(StringOnly.FIELD_CHARS));
+            // TODO Field is currently hidden, but should the field be visible in the schema
+            assertFalse(stringOnlySchema.hasField("newField"));
+            assertEquals(1, stringOnlySchema.getFieldNames().size());
+        } finally {
+            realm.close();
+        }
+    }
+
+    // Check that a Realm cannot be opened if it contain breaking schema changes, like changing a primary key
+    @Test
+    public void breakingSchemaChange_throws() {
+        SyncConfiguration config = configFactory.createSyncConfigurationBuilder(SyncTestUtils.createTestUser(), "http://foo.com/auth")
+                .schema(PrimaryKeyAsString.class)
+                .build();
+
+        // Setup initial Realm schema (with a different primary key)
+        DynamicRealm dynamicRealm = DynamicRealm.getInstance(config);
+        RealmSchema schema = dynamicRealm.getSchema();
+        dynamicRealm.beginTransaction();
+        schema.create(PrimaryKeyAsString.class.getSimpleName())
+                .addField(PrimaryKeyAsString.FIELD_PRIMARY_KEY, String.class)
+                .addField(PrimaryKeyAsString.FIELD_ID, long.class, FieldAttribute.PRIMARY_KEY);
+        dynamicRealm.commitTransaction();
+        dynamicRealm.close();
+
+        try {
+            Realm.getInstance(config);
+            fail();
+        } catch (IllegalStateException ignored) {
+        }
+    }
+
+    // Check that indexes are not being added if the schema version is the same
+    @Test
+    public void sameSchemaVersion_doNotRebuildIndexes() {
+
+        SyncConfiguration config = configFactory.createSyncConfigurationBuilder(SyncTestUtils.createTestUser(), "http://foo.com/auth")
+                .schema(IndexedFields.class)
+                .schemaVersion(42)
+                .build();
+
+        // Setup initial Realm schema (with no indexes)
+        String className = IndexedFields.class.getSimpleName();
+        DynamicRealm dynamicRealm = DynamicRealm.getInstance(config);
+        RealmSchema schema = dynamicRealm.getSchema();
+        dynamicRealm.beginTransaction();
+        schema.create(className)
+                .addField(IndexedFields.FIELD_INDEXED_STRING, String.class) // No index
+                .addField(IndexedFields.FIELD_NON_INDEXED_STRING, String.class);
+        dynamicRealm.setVersion(42);
+        dynamicRealm.commitTransaction();
+        dynamicRealm.close();
+
+        try {
+            Realm realm = Realm.getInstance(config); // Opening at same schema version (42) will not rebuild indexes
+            fail();
+        } catch (RealmMigrationNeededException ignored) {
+        }
+
+// FIXME: This is the intended behaviour
```

Comment:
That is a bit unclear at the moment and depends on the semantics that are unclear as described in : https://github.com/realm/realm-java/issues/4672

But it looks like there might be bugs in both Object Store and the annotation processor.